File: tasks_2024/day_1/day_1.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <sstream>
#include <algorithm> // For std::sort
using namespace std;


int main () 
{
	vector< int > first_column;
	vector< int > second_column;
	string line;
	string clmn_1;
	string clmn_2;
	ifstream myfile ("input.txt");
	if (myfile.is_open())
	{

		while(getline(myfile, line)) 
		{
			istringstream iss(line);
            int clmn_1, clmn_2;

            // Extract two integers from the line
            if (iss >> clmn_1 >> clmn_2) 
            {
                first_column.push_back(clmn_1);
                second_column.push_back(clmn_2);
            }
		}
		myfile.close();
	}
	else cout << "Unable to open file"; 

	// Sort the first column in ascending order
	sort(first_column.begin(), first_column.end());
	sort(second_column.begin(), second_column.end());

	int distance = 0;
	int d_helper = 0;

	for(int i = 0; i < first_column.size(); i++)
	{
		d_helper = abs(first_column[i] - second_column[i]);
		distance += d_helper;
	}
	cout << distance << endl;

	int counter = 0;
	int similarity_score = 0;
	int total_similarity_score = 0;

	for (int i = 0; i < first_column.size(); i++)
	{
		for (int j = 0; j < second_column.size(); j++)
		{
			if (first_column[i] == second_column[j])
				counter++;
		}
		similarity_score = first_column[i] * counter;
		counter = 0;
		total_similarity_score += similarity_score;
	}
	cout << "Total similarity score: " << total_similarity_score << endl;



	// //Debug print to verify the vectors
    // cout << "First Column: ";
    // for (int num : first_column) 
	// {
    //     cout << num << " ";
	// 	//cout << typeid(num).name() << " ";
	// }
    // cout << endl;

    // cout << "Second Column: ";
    // for (int num : second_column) 
    //     cout << num << " ";
    // cout << endl;


	return 0;
}

File: tasks_2024/day_1/day_1.py
# FIRST PART

# pseudo code
# put first column into array
	# sort
# put second column into array
	# sort
# iterate over each of them
# compare sorted in an increasing order values 
# increment distance

from re import T


first_column = []
second_column = []

with open("input.txt") as catalogue:
	for line in catalogue:
		column = line.split()
		fst = int(column[0])
		scnd = int(column[1])
		first_column.append(fst)
		second_column.append(scnd)

print(column)
first_column.sort()
second_column.sort()

distance = 0
d_help  = 0

for el, scnd_el in zip(first_column, second_column):
	#print(el, scnd_el)
	d_help = abs(el - scnd_el)
	#print(d_help)
	distance += d_help
	
	
#print(second_column)

print(distance)
#answer 1722302

# SECOND PART

# Test implementation
#iterate over first column (outter loop)
#iterate over second column (inner loop)
test_column_one = [3,4,2,1,3,3]
test_column_two = [4,3,5,3,9,3]

counter = 0
similarity_score = 0
total_similarity_score = 0


print(test_column_one)
for a in test_column_one:
	for b in test_column_two:
			#print(a, b)
			if a == b:
				counter += 1
				#print(counter)
	similarity_score = a * counter
	#print(f"Similarity score: {similarity_score}")
	counter = 0
	total_similarity_score += similarity_score
print(f"Total similarity score: {total_similarity_score}")



counter_pt = 0
sim_score_pt = 0
total_sim_score_pt = 0
# Implementation of the second Part
for x in first_column:
	for y in second_column:
		if x == y:
			counter_pt += 1
	sim_score_pt = x * counter_pt
	counter_pt = 0
	total_sim_score_pt += sim_score_pt
print(f"Total similarity score for the Second Part: {total_sim_score_pt}")
#answer 20373490

